C error when using an `?string` member of a struct as a parameter to pg exec function

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
V full version: V 0.4.3 5b96d8d.10d738c
OS: macos, macOS, 14.2.1, 23C71
Processor: 11 cpus, 64bit, little endian, Apple M3 Pro

vexe mtime: 2024-01-06 03:03:16

vroot: OK
VMODULES: OK
VTMP: OK

Git version: git version 2.39.3 (Apple Git-145)
Git vroot status: weekly.2024.01.test2-7-g10d738c7
.git/config present: true

CC version: Apple clang version 15.0.0 (clang-1500.1.0.2.5)
thirdparty/tcc status: thirdparty-macos-arm64 a668e5a0
```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg orm-bug.v`
```v
import db.pg

@[table: 'demo']
struct Demo {
	id int @[primary; sql: serial]
	a string
	b ?string
}

fn run() ! {
	conn := '' // insert own connection string
	db := pg.connect_with_conninfo(conn)!
	sql db { create table Demo }!
	demo := Demo{0, 'a', none}
	_ := db.exec_param_many('INSERT INTO demo(a, b) VALUES($1, $2)', [demo.a, demo.b])!
	db.close()
}

run() or { panic(err.msg()) }
```

**What did you expect to see?**

The program to compile successfully.

**What did you see instead?**
```
==================
   ^~~~~~~~~~~~~~~
/tmp/v_501/orm-bug.5278743127532159228.tmp.c:18267:197: error: initializing 'u8 *' (aka 'unsigned char *') with an expression of incompatible type '_option_string' (aka 'struct _option_string')
        _result_Array_db__pg__Row _t7 = db__pg__DB_exec_param_many(db, _SLIT("INSERT INTO demo(a, b) VALUES($1, $2)"), new_array_from_c_array(2, 2, sizeof(string), _MOV((string[2]){string_clone(demo.a), demo.b})));
                                                                                                                                                                                                           ^~~~~~
/tmp/v_501/orm-bug.5278743127532159228.tmp.c:18274:4: warning: expression result unused [-Wunused-value]
  (*(Array_db__pg__Row*)_t7.data);
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
9 warnings and 1 error generated.
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```
> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/checker/containers.v
+++ b/vlib/v/checker/containers.v
@@ -212,6 +212,20 @@ fn (mut c Checker) array_init(mut node ast.ArrayInit) ast.Type {
 				c.check_expected(typ, elem_type) or {
 					c.error('invalid array element: ${err.msg()}', expr.pos())
 				}
+				if !elem_type.has_flag(.option)
+					&& (typ.has_flag(.option) || typ.idx() == ast.none_type_idx) {
+					typ_str, elem_type_str := c.get_string_names_of(typ, elem_type)
+					if typ.idx() == ast.none_type_idx {
+						c.error('cannot use `${typ_str}` as `${elem_type_str}`', expr.pos())
+					} else {
+						c.error('cannot use `${typ_str}` as `${elem_type_str}`, it must be unwrapped first',
+							expr.pos())
+					}
+				} else if elem_type.has_flag(.option) && !typ.has_flag(.option)
+					&& typ.idx() != ast.none_type_idx && !expr.is_pure_literal() {
+					typ_str, elem_type_str := c.get_string_names_of(typ, elem_type)
+					c.error('cannot use `${typ_str}` as `${elem_type_str}`', expr.pos())
+				}
 			}
 		}
 		if node.is_fixed {

--- a/vlib/v/checker/tests/array_init_element_option_mismatch_err.out
+++ b/vlib/v/checker/tests/array_init_element_option_mismatch_err.out
@@ -1,0 +1,20 @@
+vlib/v/checker/tests/array_init_element_option_mismatch_err.vv:8:11: error: cannot use `?string` as `string`, it must be unwrapped first
+    6 | fn main() {
+    7 |     str := ?string(none)
+    8 |     _ = ['', str]
+      |              ~~~
+    9 | 
+   10 |     foo := Foo{}
+vlib/v/checker/tests/array_init_element_option_mismatch_err.vv:11:18: error: cannot use `?string` as `string`, it must be unwrapped first
+    9 | 
+   10 |     foo := Foo{}
+   11 |     _ = [foo.a, foo.b]
+      |                     ^
+   12 |     _ = [foo.b, foo.a]
+   13 | }
+vlib/v/checker/tests/array_init_element_option_mismatch_err.vv:12:18: error: cannot use `string` as `?string`
+   10 |     foo := Foo{}
+   11 |     _ = [foo.a, foo.b]
+   12 |     _ = [foo.b, foo.a]
+      |                     ^
+   13 | }

--- a/vlib/v/checker/tests/array_init_element_option_mismatch_err.vv
+++ b/vlib/v/checker/tests/array_init_element_option_mismatch_err.vv
@@ -1,0 +1,13 @@
+struct Foo {
+	a string
+	b ?string
+}
+
+fn main() {
+	str := ?string(none)
+	_ = ['', str]
+
+	foo := Foo{}
+	_ = [foo.a, foo.b]
+	_ = [foo.b, foo.a]
+}

